bolt-client/src/main.rs

```diff
@@ -0,0 +1,121 @@
+use clap::Parser;
+use eyre::{bail, Result};
+use tracing::info;
+use url::Url;
+
+mod registry;
+use registry::BoltRegistry;
+
+mod utils;
+use utils::*;
+
+#[derive(Parser)]
+struct Opts {
+    /// Bolt RPC URL to send requests to
+    #[clap(short = 'p', long, default_value = "http://135.181.191.125:8015/", env)]
+    rpc_url: Url,
+    /// Private key to sign transactions with
+    #[clap(short = 'k', long, env)]
+    private_key: String,
+    /// Optional nonce offset to use for the transaction
+    #[clap(short, long, default_value_t = 0, env)]
+    nonce_offset: u64,
+    /// Flag for generating a blob tx instead of a regular tx
+    #[clap(short, long, default_value_t = false)]
+    blob: bool,
+    /// Number of transactions to send in a sequence
+    #[clap(short, long, default_value_t = 1)]
+    count: u64,
+
+    /// Flag for using the registry to fetch the lookahead
+    #[clap(short, long, default_value_t = false)]
+    use_registry: bool,
+    /// URL of the beacon client to use for fetching the lookahead
+    /// (only used with the "use-registry" flag)
+    #[clap(short = 'b', long, env)]
+    beacon_client_url: Option<Url>,
+    /// Address of the registry contract to read bolt sidecars from
+    /// (only used with the "use-registry" flag)
+    #[clap(short, long, env, default_value = "0xdF11D829eeC4C192774F3Ec171D822f6Cb4C14d9")]
+    registry_address: Address,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt::init();
+    info!("starting bolt-client");
+
+    let _ = dotenvy::dotenv();
+    let opts = Opts::parse();
+
+    let wallet: PrivateKeySigner = opts.private_key.parse().expect("should parse private key");
+    let transaction_signer: EthereumWallet = wallet.clone().into();
+    let provider = ProviderBuilder::new().on_http(opts.rpc_url.clone());
+    let sender = wallet.address();
+
+    let (target_sidecar_url, target_slot) = if opts.use_registry {
+        // Fetch the next preconfer slot from the registry and use it
+        let beacon_api_client = BeaconApiClient::new(
+            opts.beacon_client_url.expect("beacon_client_url is required when using the registry"),
+        );
+        let registry = BoltRegistry::new(opts.rpc_url, opts.registry_address);
+        let curr_slot = get_current_slot(&beacon_api_client).await?;
+        let proposer_duties =
+            get_proposer_duties(&beacon_api_client, curr_slot, curr_slot / 32).await?;
+        match registry.next_preconfer_from_registry(proposer_duties).await {
+            Ok(Some((endpoint, slot))) => (Url::parse(&endpoint)?, slot),
+            Ok(None) => bail!("no next preconfer slot found"),
+            Err(e) => bail!("error fetching next preconfer slot from registry: {:?}", e),
+        }
+    } else {
+        // TODO: remove "cbOnly=true"
+        let url =
+            opts.rpc_url.join("proposers/lookahead?onlyActive=true&onlyFuture=true&cbOnly=true")?;
```

Comment:
presumably we're gonna have the next release ready today / tomorrow, and our partners running the sidecar won't have time to upgrade until early next week. Since this is an API breaking change, I am selecting only our proposers and avoiding others for now with this query param. this is temporary as explained in the TODO

Reply:
Makes perfect sense, thank you.